File: Java/프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.HashSet;

class Solution {
    public int[] solution(int n, String[] words) {
        int[] answer = new int[2];
        HashSet<String> usedWords = new HashSet<>(); // 사용된 단어를 저장할 집합
        
        int cnt = 1; // 사람 번호
        boolean isSuccess = true;

        for (int i = 0; i < words.length; i++) {
            // 사람 번호를 넘어가면 
            if (cnt > n) {
                cnt = 1;
            }
            
            // 끝말잇기에 실패하거나 같은 단어를 말했을 경우
            if ((i > 0 && words[i].charAt(0) != words[i-1].charAt(words[i-1].length() - 1)) 
                || usedWords.contains(words[i])) {
                
                answer[0] = cnt;  // 현재 사람 번호 저장
                answer[1] = (i / n) + 1;  // 현재 사람의 턴 계산
                isSuccess = false;
                break;
            }
            
            // 사용된 단어로 추가
            usedWords.add(words[i]);
            cnt += 1;
        }

        // 성공했다면
        if (isSuccess) {
            answer[0] = 0;
            answer[1] = 0;
        }
        
        return answer;
    }
}

File: Python3/프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.py
def solution(n, words):
    used_words = set() # 사용한 언어
    last_char = words[0][0] # 마지막 알파벳
    
    for idx, word in enumerate(words):
        # 사용했던 단어이거나 끝말이 이어지지 않은 경우
        if word in used_words or word[0] != last_char:
            player = (idx % n) + 1 # 플레이어 순서
            round_number = (idx // n) + 1 # 라운드 
            return [player, round_number]
        used_words.add(word) # 단어 사전에 등록
        last_char = word[-1] # 단어 업데이트
    
    return [0, 0]
